pkg/agent/billing.go

```diff
@@ -0,0 +1,181 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	vmapi "github.com/neondatabase/neonvm/apis/neonvm/v1"
+
+	"github.com/neondatabase/autoscaling/pkg/billing"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type BillingConfig struct {
+	URL                 string `json:"url"`
+	CollectEverySeconds uint   `json:"collectEverySeconds"`
+	PushEverySeconds    uint   `json:"pushEverySeconds"`
+	PushTimeoutSeconds  uint   `json:"pushTimeoutSeconds"`
+}
+
+type billingMetricsState struct {
+	historical map[billingMetricsKey]vmMetricsHistory
+	present    map[billingMetricsKey]vmMetrics
+	lastTime   *time.Time
+}
+
+type billingMetricsKey struct {
+	uid        types.UID
+	endpointID string
+}
+
+type vmMetricsHistory struct {
+	slices []metricsTimeSlice
+}
+
+type metricsTimeSlice struct {
+	metrics   vmMetrics
+	startTime time.Time
+	endTime   time.Time
+}
+
+type vmMetrics struct {
+	cpu uint16
+}
+
+const (
+	MetricNameCPU string = "cpu"
+	EndpointLabel string = "neon/endpoint-id"
+)
+
+func RunBillingMetricsColllector(backgroundCtx context.Context, conf *BillingConfig, store *util.WatchStore[vmapi.VirtualMachine]) error {
+	client := billing.NewClient(conf.URL, http.DefaultClient)
+
+	collectTicker := time.NewTicker(time.Second * time.Duration(conf.CollectEverySeconds))
+	defer collectTicker.Stop()
+	pushTicker := time.NewTicker(time.Second * time.Duration(conf.PushEverySeconds))
+	defer pushTicker.Stop()
+
+	state := billingMetricsState{
+		historical: make(map[billingMetricsKey]vmMetricsHistory),
+		present:    make(map[billingMetricsKey]vmMetrics),
+		lastTime:   nil,
+	}
+
+	state.collect(conf, store)
+
+	for {
+		select {
+		case <-collectTicker.C:
+			state.collect(conf, store)
```

Comment:
I was initially sort of concerned that by doing collection and pushing in the same thread, one could block the other, but I imagine that collection is always pretty fast, and if we don't collect for a couple of seconds that's probably net fine? 

The other option is to do them in different threads and use https://github.com/tychoish/fun/blob/main/pubsub/queue.go#L34 or similar to buffer/shed load.

Reply:
> I imagine that collection is always pretty fast, and if we don't collect for a couple of seconds that's probably net fine?

that's basically my thoughts as well. Setting a short timeout on pushing means we should be _mostly_ ok — but probably worth improving this eventually.

Reply:
yeah, as long as missed sending doesn't cause the payload to balloon, I think we're ok.